### ISSUE TITLE
Look for a solution to help us improving MQTT_ProcessLoop 's behaviour in user worker thread.  Thanks!

### ISSUE THREAD

**Issue by zlhk100** [aws/aws-iot-device-sdk-embedded-C#1634]
Dear AWS C SDK expert,

In our MQTT connecting and incoming message polling sequence,  we are doing something like following straightforward from one worker thread A.
      1>.  MQTT_init 
      2>  Connect to MQTT broker via mutual authN with this API.
             Openssl_Connect(mNetworkContext, &serverInfo, &opensslCredentials,            
                                        SSL_read_write_timeout_ms, SSL_read_write_timeout_ms); 
      3> Establish a clean MQTT session over authN'ed TLS with this API
             MQTT_Connect(&mMqttSessions[handle.idx]->mContext, &connectInfo, NULL, CONNACK_RECV_TIMEOUT_MS,... );
      4>. Then subscribing to some incoming topics over previous established MQTT channel persistent over entire process lifecycle.

During runtime,   from same worker thread A,   other than some other user logics (eg. user queue processing etc),   MQTT_ProcessLoop is invoked every time when worker thread A is scheduled for checking incoming MQTT messages

By our experimenting, We realized that MQTT_ProcessLoop<...> demonstrates below timing behaviour.
     /*  Calling this function blocks the calling context for a time period that
     * depends on the passed @mTimeoutMs(eg. 100ms), and the underlying transport interface SSL_read(...) implementation
     * timeouts @SSL_read_write_timeout_ms(eg. 2000ms), unless an error occurs. 
     *  **_Please blocking time will be up to MAX(mTimeoutMs, SSL_read_write_timeout_ms), which is 2000ms. _**
    status = MQTT_ProcessLoop(&mMqttSessions[handle.idx]->mContext,  mTimeoutMs); 

Here is our problem which need you guys' support to get us un-blocked for our project.   In our case, since we constantly poll the MQTT for pending messages by MQTT_ProcessLoop(...),  we expect it should return right away if nothing is pending.  Otherwise it will block our worker thread A from doing other routine operation with un-accepted latency.

Current hack we put in to fix our issue is like following:
       1> Since step 2> Openssl_Connect(...) will require sufficient timeout to get TLS hand-shaking completed, we leave @SSL_read_write_timeout_ms as is eg. (2000ms)
      2>. After TLS successfully is established,  we dynamically override underneath transport interface SSL_read<>'s timeout value to something new eg. 1ms by using raw posix socket API  like following
            setsockopt(params->socketDescriptor, SOL_SOCKET, SO_RCVTIMEO, &tm, .....)
      3> After that,  each invocation of MQTT_ProcessLoop becomes much latency friendly within our primary worker thread A.

In short,   we are looking for non-hacking solution or API for helping us to support our use case.  One option I can think of is that. eg.   AWS IOT SDK library expose an API to allow us dynamically to adjust the transport interface's timeout value like our hacking way!

Look forward to your feedbacks and much appreciated!
Lei Zhou

**Comment by zlhk100** [comment]
Do we have any updates?    Thanks,   Lei

**Comment by yourslab** [comment]
Although I don't see anything particularly wrong with what you've done, one common way is to use a blocking socket for connect (as you've already done with 2000ms). After which, you can set the socket to non-blocking. However, it may require some updates to the transport interface underneath, particularly with Openssl_send and Openssl_recv, as we don't have a mechanism yet to configure non-blocking sockets.

**Comment by zlhk100** [comment]
thanks @yourslab for the comments.  Do we plan to expose user interface to allow dynamically changing the transport I/O's blocking/non-blocking behaviour?

Regards,
Lei

**Comment by yourslab** [comment]
Thanks for making this suggestion. We will look into adding this feature into the pipeline. However, one way you can get around this limitation is by creating a wrapper function around Openssl_send and Openssl_Recv that calls `poll` or `select` on the socket before calling the aforementioned functions.

**Comment by zlhk100** [comment]
thanks @yourslab and @aggarw13 for the suggestion and we will study making proper workaround for this case.    Regards,  Lei
